### PR TITLE
DAOS-5121 Placement: Placement Changes for OSA Drain

### PR DESCRIPTION
**Pull request by PetFet**
The following changes are toward the online server drain feature.
This feature allows the draining and preemptive rebuilding of object data
including non-replicated object shards.

* Added the Drain component status
* Placement and Rebuild not support Draining targets
* Reworked tests to increase coverage
* Created tests for Drain

Signed-off-by: Peter Fetros <peter.fetros@intel.com>

### CODE REVIEW

File: src/placement/pl_map_common.c [hunk 1]
```diff
@@ -365,30 +366,14 @@ determine_valid_spares(struct pool_target *spare_tgt, struct daos_obj_md *md,
...
-		 *
-		 *      Currently, DAOS does not support the target
-		 *      re-integration. So the failure sequence for
-		 *      available spares will be the initial value
-		 *      (the oldest one). So here, we only need to
-		 *      consider those unavailable spares's failure
-		 *      sequences to find out the latest (largest).
-		 */
-		l_shard->po_fseq = f_shard->fs_fseq;
+		l_shard->po_fseq = spare_tgt->ta_comp.co_fseq;
```
Comment (NiuYawei) [inline src/placement/pl_map_common.c:369]: What's the purpose of this change? Isn't f_shared->fs_fseq == spare_tgt->ta_comp.co_fseq?

Comment (PetFet) [reply src/placement/pl_map_common.c:369]: No f_shard should be the shard we're replacing while spare_tgt is what we're replacing it with.
Without this change it appeared that sometimes when doing online drain/reintegration calculations 2 shards would end up with the same f_seq.

Comment (NiuYawei) [reply src/placement/pl_map_common.c:369]: I don't see why drain/reint needs this pl_obj_shard->po_fseq. It was used to store the "last fseq" of the shard, and DTX will use this information to choose leader.

The spare_tgt->ta_comp.co_fseq is zero in most cases, so this change will mistakenly store 0 instead of 'last fseq' in the 'po_fseq'.

BTW, the original comment is bit hard to understand, could you revise it? Thanks.

File: src/placement/jump_map.c [hunk 2]
```diff
@@ -645,14 +664,13 @@ get_object_layout(struct pl_jump_map *jmap, struct pl_obj_layout *layout,
...
-
+				state = target->ta_comp.co_status;
 				rc = remap_alloc_one(remap_list, k, target,
 						false);
 				if (rc)
 					D_GOTO(out, rc);
 
-				if (op_type == PL_PLACE_EXTENDED &&
-				   target->ta_comp.co_status == PO_COMP_ST_UP) {
+				if (can_extend(op_type, state)) {
```
Comment (gnailzenh) [inline src/placement/jump_map.c:673]: I would separately check op_type, instead of putting it in can_extend
